### PR TITLE
add GetByString

### PR DESCRIPTION
**Pull request by josharian**
For discussion.

### CODE REVIEW

**Comment by bradfitz** [comment]
Not opposed. Add an allocation AllocsPerRun test though? 

Tangentially related: we need a fast path for Gets that doesn't acquire the mutex. Could have a small type safe cache. Could even have an AfterFunc prune it over time and use atomics to occasionally bump an atime if atime is un-updated in last N seconds.

File: intern.go [hunk 1]
```diff
@@ -86,6 +86,33 @@ func Get(cmpVal interface{}) *Value {
 	return v
 }
 
+// GetString is identical to Get, except that it is specialized for strings.
+// This prevents an allocation from putting a string into an interface{}
```
Comment (danderson) [inline intern.go:90]: s/prevents/avoids/ I think?

**Review by danderson — APPROVED** [review]
LGTM. This API lets us make all IP parsing alloc-free in netaddr, which is quite nice. It also gets us closer to being at least as fast as net.ParseIP universally (we're much faster for v4 and v6, and with this change only a little bit slower on v6+zone).

**Comment by bradfitz** [comment]
GetByString maybe? Otherwise sounds like the thing you're getting is a string.

File: intern.go (hunk 1 above)
Comment (josharian) [reply intern.go]: Done, thanks.

**Comment by josharian** [comment]
> GetByString

Done.

**Review by bradfitz — COMMENTED** [review]
I think we can do this without the duplication and no allocs. Biab.

File: intern.go [hunk 2]
```diff
@@ -86,6 +86,33 @@ func Get(cmpVal interface{}) *Value {
...
+// This avoids an allocation from putting a string into an interface{}
+// to pass as an argument to Get.
+func GetByString(s string) *Value {
+	mu.Lock()
+	defer mu.Unlock()
+
+	var v *Value
+	if valSafe != nil {
+		v = valSafe[s]
+	} else if addr, ok := valMap[s]; ok {
```
Comment (bradfitz) [inline intern.go:99]: How about making a new `type key struct { s string, ei interface{} }` type, and make the maps keyed by that instead. Then Get and GetString can both call an unexported `func get` that takes that wide key (we could even add an int64 to it later).

It's pretty sad to have that gnarly subtle code duplicated.

**Pull request closed without merging**